### PR TITLE
Add picture annotation item type

### PR DESCRIPTION
**Pull request by nyalldawson**
Can render SVG or raster images as items in an annotation layer.
Options are present for:

- Locking the picture's aspect ratio
- Drawing with a background symbol
- Drawing with a border symbol
- Linked or embedded pictures
- Rendering with scale dependent sizes or fixed sizes

image

### CODE REVIEW

**Comment by nyalldawson** [comment]
Includes a whole bunch of other pre-requisite PRs which are patiently waiting review...

**Comment by nyalldawson** [comment]
@nirvn As discussed, I've now added a fixed size mode here too

File: src/core/annotations/qgsannotationpictureitem.h [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    QgsRectangle mBounds;
+
+    QSizeF mFixedSize;
+    Qgis::RenderUnit mFixedSizeUnit = Qgis::RenderUnit::Millimeters;
+
+    bool mLockAspectRatio = true;
+    bool mDrawBackground = false;
+    std::unique_ptr< QgsFillSymbol > mBackgroundSymbol;
+    bool mDrawBorder = false;
+    std::unique_ptr< QgsFillSymbol > mBorderSymbol;
```
Comment (nirvn) [inline src/core/annotations/qgsannotationpictureitem.h:250]: Would be nice to align the member name to their paired functions.

```suggestion
    bool mDrawFrame = false;
    std::unique_ptr< QgsFillSymbol > mFrameSymbol;
```

File: src/core/annotations/qgsannotationpictureitem.cpp [hunk 2]
```diff
@@ -0,0 +1,590 @@
...
+
+  element.setAttribute( QStringLiteral( "backgroundEnabled" ), mDrawBackground ? QStringLiteral( "1" ) : QStringLiteral( "0" ) );
+  if ( mBackgroundSymbol )
+  {
+    QDomElement backgroundElement = document.createElement( QStringLiteral( "backgroundSymbol" ) );
+    backgroundElement.appendChild( QgsSymbolLayerUtils::saveSymbol( QStringLiteral( "backgroundSymbol" ), mBackgroundSymbol.get(), document, context ) );
+    element.appendChild( backgroundElement );
+  }
+
+  element.setAttribute( QStringLiteral( "frameEnabled" ), mDrawBorder ? QStringLiteral( "1" ) : QStringLiteral( "0" ) );
```
Comment (nirvn) [inline src/core/annotations/qgsannotationpictureitem.cpp:219]: See my comment in the header file about Border -> Frame.

File: src/gui/annotations/qgsannotationitemguiregistry.cpp [hunk 3]
```diff
@@ -20,6 +20,10 @@
 
 #include "qgsannotationitemwidget_impl.h"
 #include "qgscreateannotationitemmaptool_impl.h"
+
+#include <QImageReader>
+#include <QFileDialog>
```
Comment (nirvn) [inline src/gui/annotations/qgsannotationitemguiregistry.cpp:25]: Is this needed?

File: src/gui/annotations/qgsannotationitemwidget_impl.cpp [hunk 4]
```diff
@@ -606,5 +609,270 @@ void QgsAnnotationLineTextItemWidget::mInsertExpressionButton_clicked()
...
+    if ( !mRadioSVG->isChecked() && QFileInfo( source ).suffix().compare( QLatin1String( "svg" ), Qt::CaseInsensitive ) == 0 )
+    {
+      mRadioSVG->setChecked( true );
+    }
+
+    onWidgetChanged();
+  } );
+
+  connect( mLockAspectRatioCheck, &QCheckBox::toggled, this, &QgsAnnotationPictureItemWidget::onWidgetChanged );
+  connect( mBorderCheckbox, &QGroupBox::toggled, this, &QgsAnnotationPictureItemWidget::onWidgetChanged );
```
Comment (nirvn) [inline src/gui/annotations/qgsannotationitemwidget_impl.cpp:658]: Border -> Frame for widget names?

**Review by nirvn — COMMENTED** [review]
Minor comments, otherwise really nice addition to annotations.

File: src/core/annotations/qgsannotationpictureitem.h (hunk 1 above)
Comment (nyalldawson) [reply src/core/annotations/qgsannotationpictureitem.h:250]: Done

File: src/gui/annotations/qgsannotationitemwidget_impl.cpp (hunk 4 above)
Comment (nyalldawson) [reply src/gui/annotations/qgsannotationitemwidget_impl.cpp:658]: Done

**Pull request merged**